### PR TITLE
Check option to see whether nested group memberships are enabled

### PR DESCRIPTION
**Pull request by mrajashree**
https://github.com/rancher/rancher/issues/14284

### CODE REVIEW

File: pkg/auth/providers/activedirectory/activedirectory_client.go [hunk 1]
```diff
@@ -123,6 +122,41 @@ func (p *adProvider) getPrincipalsFromSearchResult(result *ldapv2.SearchResult,
 	}
 	userPrincipal.Me = true
 
+	if config.NestedGroupMembershipEnabled != nil {
+		if *config.NestedGroupMembershipEnabled == false {
+			if len(memberOf) != 0 {
+				lConn, err := p.ldapConnection(config, caPool)
```
Comment (prachidamle) [inline pkg/auth/providers/activedirectory/activedirectory_client.go]: I think lConn is not used here?

File: pkg/auth/providers/activedirectory/activedirectory_client.go [hunk 2]
```diff
@@ -123,6 +122,41 @@ func (p *adProvider) getPrincipalsFromSearchResult(result *ldapv2.SearchResult,
 	}
 	userPrincipal.Me = true
 
+	if config.NestedGroupMembershipEnabled != nil {
+		if *config.NestedGroupMembershipEnabled == false {
```
Comment (prachidamle) [inline pkg/auth/providers/activedirectory/activedirectory_client.go]: Is there any reason why we are referring to the flag in different ways here? 
config.NestedGroupMembershipEnabled and *config.NestedGroupMembershipEnabled?

File: pkg/auth/providers/activedirectory/activedirectory_client.go [hunk 3]
```diff
@@ -136,14 +170,30 @@ func (p *adProvider) getPrincipalsFromSearchResult(result *ldapv2.SearchResult,
...
+	var groupPrincipals []v3.Principal
+	var nilPrincipal []v3.Principal
+
 	search := ldapv2.NewSearchRequest(searchBase,
 		ldapv2.ScopeWholeSubtree, ldapv2.NeverDerefAliases, 0, 0, false,
-		nestedGroupsQuery,
+		filter,
 		ldap.GetGroupSearchAttributes(MemberOfAttribute, ObjectClassAttribute, config), nil)
 
 	lConn, err := p.ldapConnection(config, caPool)
```
Comment (prachidamle) [inline pkg/auth/providers/activedirectory/activedirectory_client.go]: We need to close this ldap connection after the method exits

File: pkg/auth/providers/activedirectory/activedirectory_client.go (hunk 2 above)
Comment (mrajashree) [reply pkg/auth/providers/activedirectory/activedirectory_client.go]: The first to check if the pointer is not nil

File: pkg/auth/providers/activedirectory/activedirectory_client.go (hunk 3 above)
Comment (mrajashree) [reply pkg/auth/providers/activedirectory/activedirectory_client.go]: done

File: pkg/auth/providers/activedirectory/activedirectory_client.go [hunk 4]
```diff
@@ -163,18 +210,18 @@ func (p *adProvider) getPrincipalsFromSearchResult(result *ldapv2.SearchResult,
...
 			}
-			return userPrincipal, groupList, nil
+			return groupList, nil
 		}
-		return userPrincipal, groupPrincipals, err
+		return groupPrincipals, err
 	}
 
-	result, err = lConn.SearchWithPaging(search, 1000)
+	result, err := lConn.SearchWithPaging(search, 1000)
```
Comment (prachidamle) [inline pkg/auth/providers/activedirectory/activedirectory_client.go]: just curious, did we decide to use the page limit to be 1000 for nested groups?

Comment (mrajashree) [reply pkg/auth/providers/activedirectory/activedirectory_client.go]: We switch to SearchWithPaging for this issue, `https://github.com/rancher/rancher/issues/13421`, where searching for a group by name could return a large number of groups. And the nestedQuery can also return a large number of groups.

Comment (prachidamle) [reply pkg/auth/providers/activedirectory/activedirectory_client.go]: I see, thank you!

File: pkg/auth/providers/activedirectory/activedirectory_client.go [hunk 5]
```diff
@@ -163,18 +210,18 @@ func (p *adProvider) getPrincipalsFromSearchResult(result *ldapv2.SearchResult,
...
 	if err != nil {
 		// check errors
 		if ldapErr, ok := err.(*ldapv2.Error); ok && ldapErr.ResultCode == 32 {
-			return userPrincipal, groupPrincipals, err
+			return groupPrincipals, err
```
Comment (prachidamle) [inline pkg/auth/providers/activedirectory/activedirectory_client.go]: We seem to be handling ldapErr.ResultCode == 32 by returning a "Not Found" httperror elsewhere. Does this apply here? Or should we just return the err as-is?

**Comment by prachidamle** [comment]
LGTM

**Comment by cjellick** [comment]
LGTM

**Pull request merged**